### PR TITLE
Feature/deepvar recipe

### PR DESCRIPTION
**Pull request by ramprasadn**
### This PR adds | fixes:

- Program recipes, pipeline and tests for both DeepVariant and Glnexus

### How to test:

- Automatic and continuous test suite

### Expected outcome:
- [ ] Installation, unit and integration test suite pass

### Review:
- [ ] Code review
- [ ] Tests pass

This [version](https://semver.org/) is a:
- [ ] **MAJOR** - when you make incompatible API changes. If applicable record manual test results in PR header
- [x] **MINOR** - when you add functionality in a backwards compatible manner. If applicable record manual test results in PR header
- [ ] **PATCH** - when you make backwards compatible bug fixes or documentation/instructions

### CODE REVIEW

File: lib/MIP/Recipes/Analysis/Glnexus.pm [hunk 1]
```diff
@@ -0,0 +1,269 @@
...
+    );
+
+    print {$filehandle} $PIPE . $SPACE;
+
+    bcftools_view(
+        {
+            filehandle  => $filehandle,
+            infile_path => $DASH,
+        }
+    );
```
Comment (jemten) [inline lib/MIP/Recipes/Analysis/Glnexus.pm:233]: bcftools_view could output a gzipped vcf directly by adding the option `output_type => q{z}`. If you do that you could give bcftools some more threads. That being said, there might be performance benefits to piping through bcftools view without compressing to bgzip as you have written it.

File: lib/MIP/Recipes/Analysis/Glnexus.pm [hunk 2]
```diff
@@ -0,0 +1,269 @@
...
+
+    print {$filehandle} $PIPE . $SPACE;
+
+    htslib_bgzip(
+        {
+            filehandle      => $filehandle,
+            stdoutfile_path => $outfile_path,
+            write_to_stdout => 1,
```
Comment (jemten) [inline lib/MIP/Recipes/Analysis/Glnexus.pm:241]: Maybe adds some threads to the compression

**Review by jemten — CHANGES_REQUESTED** [review]
Great work Ram! You were definitely thrown into the deep end of the pool directly 😄 
I have added some suggestions and also some questions

File: lib/MIP/Recipes/Analysis/Glnexus.pm (hunk 1 above)
Comment (ramprasadn) [reply lib/MIP/Recipes/Analysis/Glnexus.pm:233]: I did not find an option to run bcftools view with multiple threads, but bgzip can do that. My guess is that's why authors of glnexus recommend using bgzip for compression. Having said that, bgzip function in MIP as it is, doesn't have that option. So, I'll create a new PR for that.

Comment (jemten) [reply lib/MIP/Recipes/Analysis/Glnexus.pm:233]: I think it's a rather new addition to [bcftools](http://samtools.github.io/bcftools/bcftools.html#view). You just add --threads to the bcftools command. Maybe there's no support for it in MIP yet. Anyway it's perfectly fine to use bgzip as well

File: lib/MIP/Program/Glnexus.pm [hunk 3]
```diff
@@ -34,20 +34,14 @@ sub glnexus_merge {
 ## Function : Perl wrapper for generic commands module
 ## Returns  : @commands
 ## Arguments: $config                 => Allows us to process vcf files from different variant callers.
-##          : $infiles_ref            => Space separated list of input vcf files
-##          : $stderrfile_path        => Stderrfile path
-##          : $stderrfile_path_append => Append stderr info to file path
-##          : $stdinfile_path         => Stdinfile path
+##          : $infile_paths_ref            => Space separated list of input vcf files
```
Comment (henrikstranneheim) [inline lib/MIP/Program/Glnexus.pm:37]: Align fat comma

File: lib/MIP/Recipes/Analysis/Deepvariant.pm [hunk 4]
```diff
@@ -0,0 +1,278 @@
...
+            required    => 1,
+            store       => \$sample_info_href,
+            strict_type => 1,
+        },
+    };
+
+    check( $tmpl, $arg_href, 1 ) or croak q{Could not parse arguments!};
+
+    use MIP::Get::File qw{ get_io_files };
+    use MIP::Get::Parameter qw{get_recipe_attributes  get_recipe_resources };
```
Comment (henrikstranneheim) [inline lib/MIP/Recipes/Analysis/Deepvariant.pm:126]:
```suggestion
    use MIP::Get::Parameter qw{ get_recipe_attributes  get_recipe_resources };
```

File: lib/MIP/Recipes/Analysis/Deepvariant.pm [hunk 5]
```diff
@@ -0,0 +1,278 @@
...
+    say {$filehandle} q{## } . $recipe_name;
+
+    deepvariant(
+        {
+            filehandle         => $filehandle,
+            infile_path        => $infile_path,
+            model_type         => uc $analysis_type,
+            num_shards         => $recipe_resource{core_number},
+            referencefile_path => $active_parameter_href->{human_genome_reference},
+            outfile_path       => $outfile_path,
```
Comment (henrikstranneheim) [inline lib/MIP/Recipes/Analysis/Deepvariant.pm:224]: Order args AB (alphabetically)

File: lib/MIP/Recipes/Analysis/Glnexus.pm [hunk 6]
```diff
@@ -0,0 +1,269 @@
...
+
+    say {$filehandle} q{## } . $recipe_name;
+
+    my $config_type = q{DeepVariant} . uc $analysis_type;
+
+    glnexus_merge(
+        {
+            filehandle       => $filehandle,
+            infile_paths_ref => \@genotype_infile_paths,
+            config           => $config_type,
```
Comment (henrikstranneheim) [inline lib/MIP/Recipes/Analysis/Glnexus.pm:222]: Order AB

File: lib/MIP/Recipes/Pipeline/Analyse_rd_dna.pm [hunk 7]
```diff
@@ -444,13 +444,15 @@ sub pipeline_analyse_rd_dna {
...
     use MIP::Recipes::Analysis::Cnvnator qw{ analysis_cnvnator };
+    use MIP::Recipes::Analysis::Deepvariant qw { analysis_deepvariant };
     use MIP::Recipes::Analysis::Delly_call qw{ analysis_delly_call };
     use MIP::Recipes::Analysis::Delly_reformat qw{ analysis_delly_reformat };
     use MIP::Recipes::Analysis::Endvariantannotationblock
       qw{ analysis_endvariantannotationblock };
     use MIP::Recipes::Analysis::Expansionhunter qw{ analysis_expansionhunter };
     use MIP::Recipes::Analysis::Fastqc qw{ analysis_fastqc };
     use MIP::Recipes::Analysis::Frequency_filter qw{ analysis_frequency_filter };
+    use MIP::Recipes::Analysis::Glnexus qw{ analysis_glnexus };
```
Comment (henrikstranneheim) [inline lib/MIP/Recipes/Pipeline/Analyse_rd_dna.pm:455]: Order AB

**Review by henrikstranneheim — APPROVED** [review]
Well done RAM!

File: lib/MIP/Recipes/Analysis/Glnexus.pm [hunk 8]
```diff
@@ -0,0 +1,269 @@
...
+            {
+                id             => $sample_id,
+                file_info_href => $file_info_href,
+                parameter_href => $parameter_href,
+                recipe_name    => $recipe_name,
+                stream         => q{in},
+            }
+        );
+        push @genotype_infile_paths, $sample_io{in}{file_path};
+        $analysis_type = $active_parameter_href->{analysis_type}{$sample_id};
```
[the remark this answers is not in the document]
Comment (ramprasadn) [reply lib/MIP/Recipes/Analysis/Glnexus.pm]: I have actually resolved this issue in the latest commit. Just forgot to mark it here as such.

File: templates/mip_install_config.yaml [hunk 9]
```diff
@@ -82,7 +82,7 @@ container:
       make_examples: /opt/deepvariant/bin/make_examples
       call_variants: /opt/deepvariant/bin/call_variants
       postprocess_variants: /opt/deepvariant/bin/postprocess_variants
-    uri: docker.io/google/deepvariant:0.10.0
+    uri: docker.io/google/deepvariant:1.0.0
```
Comment (jemten) [inline templates/mip_install_config.yaml:85]: Woho! 🥳

**Review by jemten — COMMENTED** [review]
Great work Ram! Almost there  😄 .I added som minor comments. Also there is a check failure on github actions.

**Review by jemten — APPROVED** [review]
👍

**Comment by henrikstranneheim** [comment]
@ramprasadn Great work! Looks like there was a slight drop in coverage. Seems like you should add some more test to the glnexus recipe in a future PR

**Pull request merged**